### PR TITLE
Closes #2033: Invalid FirebaseEvent

### LINKED ISSUE #2033: Crash after tapping "Clear browsing data".   (java.lang.IllegalArgumentException: Exceeding limit of param content length:1132 of 100)
STR

1. Install Rocket 2.1.1.nightly
2. Launch Rocket 2.1.1.nightly
3. Tab "Clear browsing data" function
4. The error message "Firefox Rocket Nightly keeps stopping" shows and Rocket app crash

Video: https://drive.google.com/open?id=165npwXnuMfhFenpy0ip0WduM1mqb3PMj

The Rocket 2.1.1(4587) work fine

Expected result:

- The error message does not shows up and Rocket app works fine

Actual result:

- The error message shows and Rocket app crash

Reproduce rate: 100%

Device: Redmi 4A
Rocket: 2.1.1(4696).nightly
Android: 7.1.2
Android Webview: 67.0.3396.68

### PR DESCRIPTION
**Pull request by cnevinc**
Tested in local.

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 1]
```diff
@@ -725,12 +727,19 @@ public static void showPromoteShareDialog() {
 
         EventBuilder(@NonNull String category, @NonNull String method, @Nullable String object, String value) {
             telemetryEvent = TelemetryEvent.create(category, method, object, value);
-            firebaseEvent = FirebaseEvent.create(category, method, object, value);
+            if (HomeFragment.TOPSITES_PREF.equals(value)) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Is this yet another issues other than STR described in 2033? If so, I personally prefer to separate this in another commit or even another pull request to keep a record of different things.

File: app/src/main/java/org/mozilla/focus/widget/CleanBrowsingDataPreference.java [hunk 2]
```diff
@@ -61,12 +61,12 @@ protected void onDialogClosed(boolean positiveResult) {
                 } else if (resources.getString(R.string.pref_value_clear_form_history).equals(value)) {
                     WebViewDatabase.getInstance(getContext()).clearFormData();
                 }
+                TelemetryWrapper.settingsEvent(getKey(), value);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/widget/CleanBrowsingDataPreference.java]: Make sense to me.

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java (hunk 1 above)
Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: This is the same problem when clicking "Clear browsing data". Not only the default top site creation is too long, also the four pref in "Clear browsing data" will exceed the param key limit

**Comment by cnevinc** [comment]
After discussing with Max. We will only send explicit whitelist-ed events to Firebase. We'll also rename the pref to a shorter but human readable string.  In the future, the developer who want to send pref key change event to Firebase, will need to add it to the whitelist manually. I'll add the patch later.

File: app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java [hunk 3]
```diff
@@ -96,17 +100,17 @@ public void launchByExternalAppEvent() {
     public void settingsEvent() {
         SharedPreferences pm = PreferenceManager.getDefaultSharedPreferences(InstrumentationRegistry.getTargetContext());
         for (String key : pm.getAll().keySet()) {
-            // TelemetryWrapper.settingsEvent(key, Boolean.FALSE.toString());
-            assertFirebaseEvent(TelemetryWrapper.Category.ACTION, TelemetryWrapper.Method.CHANGE, TelemetryWrapper.Object.SETTING, key, TelemetryWrapper.Extra.TO, Boolean.FALSE.toString());
+
+            TelemetryWrapper.settingsEvent(key, Boolean.FALSE.toString());
```
Comment (cnevinc) [inline app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java]: The wrapper will skill non-whitelisted keys. So I've moved the assertion for whitelisted keys in a new test below.

File: app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java [hunk 4]
```diff
@@ -96,17 +100,17 @@ public void launchByExternalAppEvent() {
...
     }
 
     @Test
     public void settingsClickEvent() {
         SharedPreferences pm = PreferenceManager.getDefaultSharedPreferences(InstrumentationRegistry.getTargetContext());
         for (String key : pm.getAll().keySet()) {
-            // TelemetryWrapper.settingsClickEvent(key);
-            assertFirebaseEvent(TelemetryWrapper.Category.ACTION, TelemetryWrapper.Method.CLICK, TelemetryWrapper.Object.SETTING, key);
+            // invalid events should be by pass
+            TelemetryWrapper.settingsClickEvent(key);
```
Comment (cnevinc) [inline app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java]: ditto

File: app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java [hunk 5]
```diff
@@ -577,5 +581,26 @@ public void showPromoteShareDialog() {
         TelemetryWrapper.showPromoteShareDialog();
     }
 
+    @Test
+    public void validPrefKeyFromWhitelistShouldPass() {
+        TelemetryWrapper.EventBuilder.lazyInit();
```
Comment (cnevinc) [inline app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java]: call lazyInit() to fill the whitelist

File: app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java [hunk 6]
```diff
@@ -577,5 +581,26 @@ public void showPromoteShareDialog() {
...
+        assertTrue(FirebaseEvent.getPrefKeyWhitelist().size() > 0);
+
+        for (String validKey : FirebaseEvent.getPrefKeyWhitelist().values()) {
+            final TelemetryWrapper.EventBuilder builder = new TelemetryWrapper.EventBuilder(TelemetryWrapper.Category.ACTION, TelemetryWrapper.Method.CHANGE, SETTING, validKey);
+            assertNotNull(builder.firebaseEvent);
+            assertNotNull(builder.telemetryEvent);
```
Comment (cnevinc) [inline app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java]: whitelist-ed pref key should worked with firebaseEvent and telemetryEvent

File: app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java [hunk 7]
```diff
@@ -577,5 +581,26 @@ public void showPromoteShareDialog() {
         TelemetryWrapper.showPromoteShareDialog();
     }
 
+    @Test
+    public void validPrefKeyFromWhitelistShouldPass() {
+        TelemetryWrapper.EventBuilder.lazyInit();
+        assertTrue(FirebaseEvent.getPrefKeyWhitelist().size() > 0);
+
```
Comment (cnevinc) [inline app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java]: make sure layInit() inits the whitelist successfully.

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 8]
```diff
@@ -253,7 +254,7 @@ public String getDefaultSearchEngineIdentifier() {
     }
 
     public static void toggleFirstRunPageEvent(boolean enableTurboMode) {
-        new EventBuilder(Category.ACTION, Method.CHANGE, Object.FIRSTRUN, Value.TURBO)
```
[the remark this answers is not in the document]
Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: fixed

File: app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java [hunk 9]
```diff
@@ -577,5 +581,26 @@ public void showPromoteShareDialog() {
...
+        for (String validKey : FirebaseEvent.getPrefKeyWhitelist().values()) {
+            final TelemetryWrapper.EventBuilder builder = new TelemetryWrapper.EventBuilder(TelemetryWrapper.Category.ACTION, TelemetryWrapper.Method.CHANGE, SETTING, validKey);
+            assertNotNull(builder.firebaseEvent);
+            assertNotNull(builder.telemetryEvent);
+        }
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidPrefKeyShouldThrowException() {
```
[the remark this answers is not in the document]
Comment (cnevinc) [reply app/src/androidTest/java/org/mozilla/focus/telemetry/TelemetryWrapperTest.java]: will remove it

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java [hunk 10]
```diff
@@ -132,24 +132,29 @@ public void setParam(Bundle bundle) {
...
     }
 
-    private static boolean isValueInWhiteList(@NonNull String value) {
-        return existingPreferenceKey.contains(value);
+    static String getValidPrefKey(@NonNull String value) {
+        return prefKeyWhitelist.get(value);
+    }
+
+    @VisibleForTesting
+    static Map<String, String> getPrefKeyWhitelist() {
```
[the remark this answers is not in the document]
Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: okay

**Pull request merged**